sw/device/tests/dif/dif_keymgr_unittest.cc

```diff
@@ -0,0 +1,1114 @@
+        break;
+    }
+  }
+};
+
+TEST_F(AdvanceStateTest, BadArgsNoKeymgr) {
+  EXPECT_EQ(dif_keymgr_advance_state(nullptr, &kStateParams),
+            kDifKeymgrLockableBadArg);
+  EXPECT_EQ(dif_keymgr_advance_state(nullptr, nullptr),
+            kDifKeymgrLockableBadArg);
+}
+
+class AdvanceToOperational : public AdvanceStateTest,
+                             public testing::WithParamInterface<uint32_t> {};
+
+TEST_P(AdvanceToOperational, BadArgsToOperationalWithoutParams) {
+  ExpectIdleAtState(GetParam());
+
+  EXPECT_EQ(dif_keymgr_advance_state(&keymgr_, nullptr),
+            kDifKeymgrLockableBadArg);
+}
+
+INSTANTIATE_TEST_SUITE_P(AdvanceToOperational, AdvanceToOperational,
+                         testing::ValuesIn(kStatesWithOperationalNextStates));
+
+class AdvanceToNonOperational : public AdvanceStateTest,
+                                public testing::WithParamInterface<uint32_t> {};
+
+TEST_P(AdvanceToNonOperational, BadArgsToNonOperationalWithParams) {
+  ExpectIdleAtState(GetParam());
+
+  EXPECT_EQ(dif_keymgr_advance_state(&keymgr_, &kStateParams),
+            kDifKeymgrLockableBadArg);
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    AdvanceToNonOperational, AdvanceToNonOperational,
+    testing::ValuesIn(kStatesWithNonOperationalNextStates));
+
+TEST_F(AdvanceStateTest, LockedBusy) {
+  ExpectBusy();
+
+  EXPECT_EQ(dif_keymgr_advance_state(&keymgr_, &kStateParams),
+            kDifKeymgrLockableLocked);
+}
+
+TEST_F(AdvanceStateTest, LockedConfig) {
+  ExpectLockedConfig();
+
+  EXPECT_EQ(dif_keymgr_advance_state(&keymgr_, &kStateParams),
+            kDifKeymgrLockableLocked);
+}
+
+TEST_P(AdvanceToOperational, LockedBinding) {
+  ExpectIdleAtState(GetParam());
+  EXPECT_READ32(KEYMGR_SW_BINDING_REGWEN_REG_OFFSET, 0);
+
+  EXPECT_EQ(dif_keymgr_advance_state(&keymgr_, &kStateParams),
+            kDifKeymgrLockableLocked);
+}
+
+TEST_P(AdvanceToOperational, LockedMaxKeyVersion) {
+  ExpectIdleAtState(GetParam());
+  EXPECT_READ32(KEYMGR_SW_BINDING_REGWEN_REG_OFFSET,
+                {{
+                    .offset = KEYMGR_SW_BINDING_REGWEN_EN_BIT,
+                    .value = 1,
+                }});
+  EXPECT_READ32(GetMaxKeyVersionRegInfo(GetParam()).wen_offset, 0);
+
+  EXPECT_EQ(dif_keymgr_advance_state(&keymgr_, &kStateParams),
+            kDifKeymgrLockableLocked);
+}
+
+TEST_P(AdvanceToOperational, Success) {
+  ExpectIdleAtState(GetParam());
+  EXPECT_READ32(KEYMGR_SW_BINDING_REGWEN_REG_OFFSET,
+                {{
+                    .offset = KEYMGR_SW_BINDING_REGWEN_EN_BIT,
+                    .value = 1,
+                }});
+  auto reg_info = GetMaxKeyVersionRegInfo(GetParam());
+  EXPECT_READ32(reg_info.wen_offset, {{
+                                         .offset = reg_info.wen_bit_index,
+                                         .value = 1,
+                                     }});
+  size_t binding_len = sizeof(kStateParams.binding_value) /
+                       sizeof(kStateParams.binding_value[0]);
```

Comment:
You *could* pull in memory.h and use ARRAYSIZE...

Reply:
Tried, but a bunch of declarations in memory.h conflict with the ones in string.h that google test pulls in...